File: courses/codes/SSSP.java
import java.util.*;

public class SSSP {
	
	public static Map<String, Integer> dijkstra(Map<String, List<String>> A, Map<String, Map<String, Integer>> W, String s) {
        Map<String, Integer> dist = new HashMap<>();
        dist.put(s, 0);
        for (String v : A.keySet()) {
            dist.put(v, Integer.MAX_VALUE);
        }
        dist.put(s, 0);
        Set<String> S = new HashSet<>();
        List<String> Q = new ArrayList<>(A.keySet());

        while (!Q.isEmpty()) {

            int min_val = Integer.MAX_VALUE;
            String u = null;
            for (String node : Q) {
                if (dist.get(node) < min_val) {
                    min_val = dist.get(node);
                    u = node;
                }
            }

            Q.remove(u);
            S.add(u);

            for (String v : A.get(u)) {
                if (S.contains(v)) {
                    continue;
                }
                int weight = W.get(u).getOrDefault(v, Integer.MAX_VALUE);
                if (dist.get(v) > dist.get(u) + weight) {
                    dist.put(v, dist.get(u) + weight);
                }
            }
        }

        return dist;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int P = scanner.nextInt();
        Map<String, List<String>> G = new HashMap<>();
        Map<String, Map<String, Integer>> W = new HashMap<>();
        Set<String> nodes = new HashSet<>();

        for (int i = 0; i < P; i++) {
            String s = scanner.next();
            String d = scanner.next();
            int c = scanner.nextInt();

            nodes.add(s);
            nodes.add(d);
            if (!G.containsKey(s)) {
                G.put(s, new ArrayList<>());
            }
            G.get(s).add(d);
            if (!W.containsKey(s)) {
                W.put(s, new HashMap<>());
            }
            W.get(s).put(d, c);
        }

        String home = scanner.next();
        String dest = scanner.next();

        Map<String, Integer> sssp = dijkstra(G, W, home);
        System.out.println(sssp.get(dest));
    }
}

File: courses/codes/sssp.py
def dijkstra(A, W, s):
    dist = dict()
    dist[s] = 0
    for v in A.keys():
        dist[v] = float("inf")
    dist[s] = 0
    S = set()
    Q = list(A.keys())

    while len(Q) > 0:

        min_val = min([dist[node] for node in Q])
        u = [key for key, value in dist.items() if value == min_val and key in Q][0]

        Q.remove(u)
        S.add(u)

        for v in A[u]:
            if v in S:
                continue
            if dist[v] > dist[u] + W[(u, v)]:
                dist[v] = dist[u] + W[(u, v)]

    return dist


P = int(input())
G = dict()
W = dict()
nodes = set()
edges = set()
for _ in range(P):
    arr = input().split(" ")
    s, d, c = arr[0], arr[1], int(arr[2])
    nodes.add(s)
    nodes.add(d)
    edge = (s,d)
    edges.add(edge)
    W[edge] = c

for node in nodes:
    if node not in G:
        G[node] = []

for edge in edges:
    e1, e2 = edge
    G[e1].append(e2)


home, dest = input().split(" ")

sssp = dijkstra(G, W, home)
print(sssp[dest]) 